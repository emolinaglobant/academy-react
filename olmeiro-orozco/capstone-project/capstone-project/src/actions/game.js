import { getQuestions } from "../helpers/getQuestions";
import { types } from "../types/types";

export const startRound = (start, numberCategory, difficulty) => {
  return {
    type: types.startRound,
    payload: {
      start: start,
      numberCategory: numberCategory,
      difficulty: difficulty,
    },
  };
};

export const loadQuestions = () => {
  return async (dispatch, getState) => {
    const { numberCategory, difficulty } = getState().game;
    const data = await getQuestions(numberCategory, difficulty);

    dispatch(storeQuestions(data));
  };
};

export const storeQuestions = (questions) => {
  return {
    type: types.gameStoreQuestions,
    payload: {
      questions,
    },
  };
};

export const increasePrize = (increase) => {
  return {
    type: types.gameIncreasePrize,
    payload: {
      totalPrize: increase,
    },
  };
};

export const addRound = (increase) => {
  return {
    type: types.gameAddRound,
    payload: {
      round: increase,
    },
  };
};

export const roundChange = (value) => {
  return {
    type: types.gameChangeRound,
    payload: {
      changeRound: value,
    },
  };
};

export const resetGame = () => {
  return {
    type: types.gameReset,
  };
};
