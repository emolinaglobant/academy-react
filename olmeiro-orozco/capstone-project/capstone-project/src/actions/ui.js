import { types } from "../types/types";

export const setError = (err) => {
  return {
    type: types.uiSetError,
    payload: err,
  };
};

export const removeError = (err) => {
  return {
    type: types.uiRemoveError,
  };
};

export const startLoading = () => ({
  type: types.uiStartLoading,
});

export const finishLoading = () => ({
  type: types.uiFinishLoading,
});
