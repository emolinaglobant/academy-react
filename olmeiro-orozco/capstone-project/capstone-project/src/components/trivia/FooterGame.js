import React from "react";

export const FooterGame = () => {
  return (
    <footer className="trivia__footer">
      <p>all rights reserved - by @olmeiro_orozco 2022</p>
    </footer>
  );
};