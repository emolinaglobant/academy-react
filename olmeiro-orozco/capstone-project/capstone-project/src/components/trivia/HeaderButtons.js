import React from 'react'

export const HeaderButtons = () => {
  return (
     <>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__warning btn__header-question">1</button>
          <button type="button" className="btn btn__warning btn__header-question">2</button>
          <button type="button" className="btn btn__warning btn__header-question">3</button>
          <button type="button" className="btn btn__warning btn__header-question">4</button>
          <button type="button" className="btn btn__warning btn__header-question">5</button>
        </div>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__warning btn__header-question">6</button>
          <button type="button" className="btn btn__warning btn__header-question">7</button>
          <button type="button" className="btn btn__warning btn__header-question">8</button>
          <button type="button" className="btn btn__warning btn__header-question">9</button>
          <button type="button" className="btn btn__warning btn__header-question">10</button>
        </div>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__success btn__header-question">11</button>
          <button type="button" className="btn btn__success btn__header-question">12</button>
          <button type="button" className="btn btn__success btn__header-question">13</button>
          <button type="button" className="btn btn__success btn__header-question">14</button>
          <button type="button" className="btn btn__success btn__header-question">15</button>
        </div>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__success btn__header-question">16</button>
          <button type="button" className="btn btn__success btn__header-question">17</button>
          <button type="button" className="btn btn__success btn__header-question">18</button>
          <button type="button" className="btn btn__success btn__header-question">19</button>
          <button type="button" className="btn btn__success btn__header-question">20</button>
        </div>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__danger btn__header-question">21</button>
          <button type="button" className="btn btn__danger btn__header-question">22</button>
          <button type="button" className="btn btn__danger btn__header-question">23</button>
          <button type="button" className="btn btn__danger btn__header-question">24</button>
          <button type="button" className="btn btn__danger btn__header-question">25</button>
        </div>
     </>
  )
}