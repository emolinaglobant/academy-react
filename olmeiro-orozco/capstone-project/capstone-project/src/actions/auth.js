import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from "firebase/auth";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import Swal from "sweetalert2";

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    const auth = getAuth();
    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));

        dispatch(finishLoading());
      })
      .catch((e) => {
        dispatch(finishLoading());
        Swal.fire({
          icon: 'error',
          title: 'Oops...',
          text: 'Something went wrong!',
          footer: 'Please confirm your Email and Password.'
        })
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    const auth = getAuth();
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });

        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        console.log(error);
        Swal.fire("Error", error.message, "error");
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => {
  return {
    type: types.login,
    payload: {
      uid,
      displayName,
    },
  };
};

export const startLogout = () => {
  return async (dispath) => {
    const auth = getAuth();
    await signOut(auth);

    dispath(logout());
  };
};

export const logout = () => {
  return {
    type: types.logout,
  };
};
