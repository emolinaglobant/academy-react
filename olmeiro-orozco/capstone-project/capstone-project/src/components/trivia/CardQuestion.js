import { React, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import Swal from "sweetalert2";

import { increasePrize, addRound, resetGame, roundChange } from "../../actions/game";
import { getRandom } from "../../helpers/getRandom";

export const CardQuestion = () => {
  const allCategories = document.querySelectorAll(".category");

  const [numberQuestionAsked, setNumberQuestionAsked] = useState(0);
  const [questionIndex, setQuestionIndex] = useState(0);
  const [prize, setPrize] = useState(1);
  const [round, setRound] = useState(1);
  const [startGame, setStart] = useState(false);
  const [disabled, setDisabled] = useState(false);

  const dispatch = useDispatch();
  const { questions, start, changeRound } = useSelector((state) => state.game);
  
  useEffect(() => {
    setStart(start);
    setDisabled(changeRound);
    dispatch(addRound(round));
  }, [questions, round, prize, dispatch, changeRound]);

  const checkAnswer = (e) => {
    let answer = e.target.dataset.answer;
    if (answer === "correct_answer") {
      if (prize % 5 === 0 && numberQuestionAsked !== 1 && prize !== 25) {
        setNumberQuestionAsked(1);
        setQuestionIndex(0);
        setPrize(prize + 1);
        setRound(round + 1);

        setDisabled(true);
        dispatch(roundChange(true));

        dispatch(increasePrize(prize));
        dispatch(addRound(round));

        Swal.fire({
          icon: "success",
          title: "Excelent...",
          text: "Please choose another category",
        });
      } else if (prize % 5 === 0 && prize === 25) {
        const imagePath = "/assets/trivia2.jpg";
        Swal.fire({
          title: 'Triviaaa!',
          text: 'You are amazing.',
          imageUrl: `${imagePath}`,
          imageWidth: 400,
          imageHeight: 200,
          imageAlt: 'Custom image',
        })
        setStart(false);
        setQuestionIndex(0);
        setNumberQuestionAsked(1);
        setPrize(1);

        allCategories.forEach((btn) => {
          btn.removeAttribute("disabled");
        });
        dispatch(resetGame());
      } else {
        
        setNumberQuestionAsked(numberQuestionAsked + 1);
        setQuestionIndex(questionIndex + 1);
        setPrize(prize + 1);
        dispatch(increasePrize(prize));

        Swal.fire({
          icon: "success",
          title: "Excellent...",
          text: "Keep playing, click on the next question.",
        });
      }
    } else {
      handleErrorQuestion();
    }
  };

  const handleErrorQuestion = () => {
    Swal.fire({
      icon: "error",
      title: "Oops...",
      text: "Nice try!",
      footer: 'Try it again, you can do it!',
    });
    setStart(false);
    setQuestionIndex(0);
    setNumberQuestionAsked(1);
    setPrize(1);

    allCategories.forEach((btn) => {
      btn.removeAttribute("disabled");
    });
    dispatch(resetGame());
  };

  const handleRetire = () => {
    Swal.fire({
      icon: "error",
      title: "You won accumulated!",
      text: "Try it again!",
      footer: "Choose another category",
    });
    setStart(false);
    setQuestionIndex(0);
    setNumberQuestionAsked(1);
    setPrize(1);

    dispatch(resetGame());

    allCategories.forEach((btn) => {
      btn.removeAttribute("disabled");
    });
  };

  const order = getRandom(0, 5);
  return (
    <>
      {startGame && (
        <article className="card__question">
          <header className="card__question-title">
            <h3>
              {questions === undefined
                ? "loading"
                : questions[questionIndex].question}
            </h3>
            <h5>choose one option, only one is true.</h5>
          </header>

          <section className="card__question-options">
            <section className="card__options">
              <section className={`card__option order-${order}`}>
                <button
                  type="button"
                  data-answer="correct_answer"
                  className="btn btn__option btn-answer btn-option"
                  disabled={disabled}
                  onClick={checkAnswer}
                >
                  {questions === undefined
                    ? "Loading"
                    : questions[questionIndex].correct_answer}
                </button>
              </section>

              <section className="card__option order-1">
                <button
                  type="button"
                  data-answer="incorrect_answer"
                  className="btn btn__option btn-answer btn-option"
                  disabled={disabled}
                  onClick={checkAnswer}
                >
                  {questions === undefined
                    ? "Loading"
                    : questions[questionIndex].incorrect_answers[0]}
                </button>
              </section>

              <section className="card__option order-2">
                <button
                  type="button"
                  data-answer="incorrect_answer"
                  className="btn btn__option btn-answer btn-option"
                  disabled={disabled}
                  onClick={checkAnswer}
                >
                  {questions === undefined
                    ? "Loading"
                    : questions[questionIndex].incorrect_answers[1]}
                </button>
              </section>

              <section className="card__option order-3">
                <button
                  type="button"
                  data-answer="incorrect_answer"
                  className="btn btn__option btn-answer btn-option"
                  disabled={disabled}
                  onClick={checkAnswer}
                >
                  {questions === undefined
                    ? "Loading"
                    : questions[questionIndex].incorrect_answers[2]}
                </button>
              </section>
            </section>
            <footer className="card__footer">
              <button
                type="button"
                className="btn btn__retirement"
                id="retire"
                onClick={handleRetire}
              >
                I want to retire
              </button>
            </footer>
          </section>
        </article>
      )}

      <section
        className="card__start"
        style={{ display: start ? "none" : "flex" }}
      >
        <h1>Please choose a category to start.</h1>
      </section>
    </>
  );
};