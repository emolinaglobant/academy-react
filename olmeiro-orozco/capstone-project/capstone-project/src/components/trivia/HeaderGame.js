import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { startLogout } from "../../actions/auth";
import { HeaderButtons } from "./HeaderButtons";

export const HeaderGame = () => {
  const { name } = useSelector((state) => state.auth);

  const imagePath = "/assets/trivia2.jpg";

  const dispatch = useDispatch();

  const handleLogout = () => {
    dispatch(startLogout());
  };

  return (
    <>
      <header className="trivia__header">
        <img
          src={imagePath}
          className="trivia__img animate__animated animate__flash"
          alt="trivia game"
        />
        <h1 className="animate__animated animate__bounce">
          Trivia Game: {name}
        </h1>
        <Link to="/rules" className="trivia__header-link">
          Rules
        </Link>
        <button
          className="btn btn__end animate__animated animate__rubberBand"
          onClick={handleLogout}
        >
          Logout
        </button>
      </header>
      <section className="trivia__header-buttons">
        <HeaderButtons />
      </section>
    </>
  );
};
