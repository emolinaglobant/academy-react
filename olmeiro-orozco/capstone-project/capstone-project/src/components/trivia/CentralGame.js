import React from "react";
import { CardQuestion } from "./CardQuestion";

export const CentralGame = () => {
  return (
    <section className="trivia__central animate__animated animate__backInUp">
      <CardQuestion />
    </section>
  );
};