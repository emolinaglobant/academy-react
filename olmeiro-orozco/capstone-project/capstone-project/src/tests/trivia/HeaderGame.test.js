import React from "react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import { mount } from "enzyme";
import { HeaderGame } from "../../components/trivia/HeaderGame";
import { startLogout } from "../../actions/game";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

jest.mock("../../actions/game", () => ({
  startLogout: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    ui: "UIeyCuTetnUjy4TAogQuh2eWMI63",
    name: "testing",
  },
  ui: {
    loading: false,
    msgError: null,
  },
  game: {
    changeRound: false,
    difficulty: "easy",
    numberCategory: 0,
    questions: [],
    round: 1,
    start: false,
    totalPrize: 0,
  },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <HeaderGame />
    </MemoryRouter>
  </Provider>
);

describe("Tests component <HeaderGame />", () => {
  test("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should show name of user testing', () => { 
    const nameUser = wrapper.find('h1');
    expect(nameUser.text()).toEqual("Trivia Game: testing");
   });

   test('should dispatch the action startLogout', () => { 
      wrapper.find('.btn__end').prop('onClick')();
      expect(startLogout).toBeCalledTimes(0);
    })
});
