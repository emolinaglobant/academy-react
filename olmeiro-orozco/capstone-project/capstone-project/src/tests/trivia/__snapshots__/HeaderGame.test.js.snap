// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests component <HeaderGame /> should render correctly 1`] = `
Array [
  <header
    className="trivia__header"
  >
    <img
      alt="trivia game"
      className="trivia__img animate__animated animate__flash"
      src="/assets/trivia2.jpg"
    />
    <h1
      className="animate__animated animate__bounce"
    >
      Trivia Game: 
      testing
    </h1>
    <a
      className="trivia__header-link"
      href="/rules"
      onClick={[Function]}
    >
      Rules
    </a>
    <button
      className="btn btn__end animate__animated animate__rubberBand"
      onClick={[Function]}
    >
      Logout
    </button>
  </header>,
  <section
    className="trivia__header-buttons"
  >
    Array [
      <div
        className="btn__header-warginbox"
      >
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          1
        </button>
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          2
        </button>
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          3
        </button>
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          4
        </button>
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          5
        </button>
      </div>,
      <div
        className="btn__header-warginbox"
      >
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          6
        </button>
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          7
        </button>
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          8
        </button>
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          9
        </button>
        <button
          className="btn btn__warning btn__header-question"
          type="button"
        >
          10
        </button>
      </div>,
      <div
        className="btn__header-warginbox"
      >
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          11
        </button>
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          12
        </button>
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          13
        </button>
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          14
        </button>
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          15
        </button>
      </div>,
      <div
        className="btn__header-warginbox"
      >
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          16
        </button>
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          17
        </button>
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          18
        </button>
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          19
        </button>
        <button
          className="btn btn__success btn__header-question"
          type="button"
        >
          20
        </button>
      </div>,
      <div
        className="btn__header-warginbox"
      >
        <button
          className="btn btn__danger btn__header-question"
          type="button"
        >
          21
        </button>
        <button
          className="btn btn__danger btn__header-question"
          type="button"
        >
          22
        </button>
        <button
          className="btn btn__danger btn__header-question"
          type="button"
        >
          23
        </button>
        <button
          className="btn btn__danger btn__header-question"
          type="button"
        >
          24
        </button>
        <button
          className="btn btn__danger btn__header-question"
          type="button"
        >
          25
        </button>
      </div>,
    ]
  </section>,
]
`;
